pageserver/src/storage_sync2.rs

```diff
@@ -649,6 +689,25 @@ impl RemoteTimelineClient {
         let metadata_bytes = upload_queue.latest_metadata.to_bytes()?;
         let disk_consistent_lsn = upload_queue.latest_metadata.disk_consistent_lsn();
 
+        for name in names {
+            if !upload_queue.latest_files.contains_key(name) {
+                anyhow::bail!(
+                    "name {} not found in upload_queue.latest_files",
+                    name.file_name()
+                );
+            }
+        }
+        let mut delete_ops = Vec::with_capacity(names.len());
+        for name in names {
+            let metadata = upload_queue
+                .latest_files
+                .remove(name)
+                // XXX assert uniqueness of names
+                .expect("we just checked above");
+            let op = UploadOp::Delete(RemoteOpFileKind::Layer, name.clone(), metadata);
+            delete_ops.push(op);
+        }
```

Comment:
There's a comment  below about "Once we start removing files from upload_queue.latest_files ..." that is a bit out of place now

Reply:
Pushed down the code where we do the .remove() calls into the no_bail_here closure.
It's a little further apart from the place where we do the check, but it's better to keep it in the closure, I'd say.